lib/netext/dialer.go

```diff
@@ -33,34 +35,96 @@ import (
 
 	"github.com/loadimpact/k6/lib"
 	"github.com/loadimpact/k6/lib/metrics"
+	"github.com/loadimpact/k6/lib/types"
 	"github.com/loadimpact/k6/stats"
 )
 
-// dnsResolver is an interface that fetches dns information
-// about a given address.
-type dnsResolver interface {
-	FetchOne(address string) (net.IP, error)
+// DNSResolver is an interface that fetches DNS information about a given host.
+type DNSResolver interface {
+	Fetch(host string) (net.IP, error)
+}
+
+type resolver struct {
+	cache      *dnscache.Resolver
+	strategy   lib.DNSStrategy
+	m          sync.Mutex
+	roundRobin map[string]uint8
+}
+
+// NewDNSResolver returns a new DNS resolver. If ttl is a valid duration,
+// responses will be cached for the specified period, or indefinitely if it's 0.
+// The IP returned from Fetch() will be selected based on the given strategy.
+func NewDNSResolver(ttl types.NullDuration, strategy lib.DNSStrategy) DNSResolver {
+	var cache *dnscache.Resolver
+	if ttl.Valid {
+		cache = dnscache.New(time.Duration(ttl.Duration))
+	}
+	return &resolver{
+		cache:      cache,
+		strategy:   strategy,
+		roundRobin: make(map[string]uint8),
+	}
+}
+
+// Fetch does a DNS lookup or uses the cache if enabled, and returns a single IP
+// selected with the preset strategy.
+func (r *resolver) Fetch(host string) (net.IP, error) {
+	var (
+		ips []net.IP
+		err error
+	)
+	if r.cache != nil {
+		ips, err = r.cache.Fetch(host)
+	} else {
+		ips, err = net.LookupIP(host)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if len(ips) == 0 {
+		return nil, errors.Errorf("DNS lookup for %s returned 0 IPs", host)
+	}
+
+	return r.selectOne(host, ips), nil
+}
+
+func (r *resolver) selectOne(host string, ips []net.IP) net.IP {
+	switch r.strategy {
+	case lib.DNSFirst:
+		return ips[0]
+	case lib.DNSRoundRobin:
+		r.m.Lock()
+		defer func() {
+			r.roundRobin[host]++
+			r.m.Unlock()
+		}()
+		// TODO: The index approach is not stable and we'll need to track
+		// individual IPs... Do we really want to implement an additional cache
+		// for this, since dnscache does not expose its internal one? If so, we
+		// should scrap dnscache entirely and re-implement it here.
+		return ips[int(r.roundRobin[host])%len(ips)]
```

Comment:
Given the size of the dnscache ... code ... I am all for just implementing our own

Reply:
In the latest changes I re-implemented the caching with an approach I discussed with Ned, so `viki-org/dnscache` is officially gone 🎉 . It doesn't use a goroutine to update all hosts when the TTL expires globally, and instead keeps track of the last lookup time per host and refreshes when that is over the TTL. There's no cache eviction, but I'm not sure it's needed since it wasn't done before either. Let me know.